streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -623,63 +642,75 @@ void runOnce() {
+                    // so we can just avoid the method call when we didn't process anything.
+                    processRateSensor.record(processed, now);
+
+                    // This metric is scaled to represent the _average_ processing time of _each_
+                    // task. Note, it's hard to interpret this as defined, but we would need a KIP
+                    // to change it to simply report the overall time spent processing all tasks.
+                    final long processLatency = advanceNowAndComputeLatency();
+                    processLatencySensor.record(processLatency / (double) processed, now);
+                    totalProcessLatency += processLatency;
+                }
+
+                final int punctuated = taskManager.punctuate();
+                if (punctuated > 0) {
+                    final long punctuateLatency = advanceNowAndComputeLatency();
+                    punctuateSensor.record(punctuateLatency / (double) punctuated, now);
+                    totalPunctuateLatency += punctuateLatency;
                 }
 
-                timeSinceLastPoll = Math.max(now - lastPollMs, 0);
+                final int committed = maybeCommit();
+                if (committed > 0) {
+                    final long commitLatency = advanceNowAndComputeLatency();
+                    commitSensor.record(commitLatency / (double) committed, now);
 
-                if (maybePunctuate() || maybeCommit()) {
-                    numIterations = numIterations > 1 ? numIterations / 2 : numIterations;
-                } else if (timeSinceLastPoll > maxPollTimeMs / 2) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Committed all active tasks {} and standby tasks {} in {}ms",
+                            taskManager.activeTaskIds(), taskManager.standbyTaskIds(), commitLatency);
+                    }
+
+                    totalCommitLatency += commitLatency;
+                }
+
+                if (processed == 0) {
+                    // if there is no records to be processed, exit after punctuate / commit
+                    break;
+                } else if (Math.max(now - lastPollMs, 0) > maxPollTimeMs / 2) {
                     numIterations = numIterations > 1 ? numIterations / 2 : numIterations;
                     break;
-                } else if (processed > 0) {
+                } else if (punctuated > 0 || committed > 0) {
+                    numIterations = numIterations > 1 ? numIterations / 2 : numIterations;
+                } else {
                     numIterations++;
                 }
-            } while (processed > 0);
-
-            maybeCommit();
+            } while (true);
         }
+
+        final long runOnceLatency = pollLatency + restoreLatency + totalCommitLatency + totalProcessLatency + totalPunctuateLatency;
+        processRatioSensor.record((double) totalProcessLatency / runOnceLatency);
```

Comment:
Note I did not record sensor  for restoration; it is intentional since we will be moving that out of the thread here. Atm the restoration ratio can be inferred by 100% - other four.

Reply:
Here's a few considerations: 1) when we move that to another thread(s), then the restoration-ratio would basically be 100% of that thread(s), unless we separate the the restoreConsumer.poll call out of applying updates here, but in that case we'd probably introduce with different tags, 2) if we are moving it to a thread pool instead of a single thread, then there would be no `thread-id` in the tags, but some other ids.

So I decided to defer adding this metric until we've figured out exactly how we are going to restore in the other threads --- I'm not suggesting that we never add this metric, just defer it :)

Reply:
Ok, sounds good.